### PR TITLE
Modal Animation Types

### PR DESCRIPTION
**Pull request by jemise111**
Currently the Modal component uses the slide up / down animation for presenting and hiding the Modal with no options. This PR gives users a choice to use a fade in / out animation or the current slide animation (slide is the default). Android and iOS.

  

I've updated the UIExplorer and documentation.

image

Thanks!

### CODE REVIEW

**Comment by dmmiller** [comment]
The android stuff looks good.  @javache do you want to look at the iOS stuff?

File: Examples/UIExplorer/ModalExample.js [hunk 1]
```diff
@@ -117,6 +126,19 @@ var ModalExample = React.createClass({
           <Switch value={this.state.animated} onValueChange={this._toggleAnimated} />
         </View>
 
+        {
+          this.state.animated &&
+	        <View style={styles.row}>
+	          <Text style={styles.rowTitle}>Animtion Type</Text>
```
Comment (charpeni) [inline Examples/UIExplorer/ModalExample.js]: Typo

**Comment by javache** [comment]
Can we simplify the API here (while keeping the old API backwards compatible) and have a property: animationType: 'none' | 'slide' | 'fade'?

Other than that, the code looks good.

**Comment by jemise111** [comment]
@javache good suggestion, done!

File: Libraries/Modal/Modal.js [hunk 2]
```diff
@@ -58,6 +59,7 @@ class Modal extends React.Component {
     return (
       <RCTModalHostView
         animated={this.props.animated}
```
Comment (javache) [inline Libraries/Modal/Modal.js]: Can we keep the backwards-compatible check for animated in the JS here. That way we don't to maintain it twice in native. If animationType is undefined and animated is set, set animationType to slide.

Comment (grabbou) [inline Libraries/Modal/Modal.js]: Also `console.warn` would be nice to have here, e.g. "animated={true} prop is deprecated, use animationType="slide" instead"

File: Libraries/Modal/Modal.js [hunk 3]
```diff
@@ -77,6 +79,19 @@ class Modal extends React.Component {
   }
 }
 
+Modal.propTypes = {
+  animationType: PropTypes.oneOf(['none', 'slide', 'fade']),
```
Comment (grabbou) [inline Libraries/Modal/Modal.js]: Wouldn't it be cool to export these constants as statics on Modal, so we can `Modal.AnimationType.Slide` and `Modal.AnimationType.Fade` etc ?

Comment (satya164) [inline Libraries/Modal/Modal.js]: @grabbou I've seen RN core team prefer enums over exported constants.

Comment (grabbou) [inline Libraries/Modal/Modal.js]: Ah, cool - will keep in mind for the future, thanks :)

File: Libraries/Modal/Modal.js (hunk 2 above)
Comment (jemise111) [inline Libraries/Modal/Modal.js]: @javache @grabbou done! Thanks for the suggestions

File: Libraries/Modal/Modal.js [hunk 4]
```diff
@@ -35,15 +36,19 @@ const RCTModalHostView = requireNativeComponent('RCTModalHostView', null);
...
+    animationType: PropTypes.oneOf(['none', 'slide', 'fade']),
     transparent: PropTypes.bool,
     visible: PropTypes.bool,
     onRequestClose: Platform.OS === 'android' ? PropTypes.func.isRequired : PropTypes.func,
     onShow: PropTypes.func,
   };
 
   static defaultProps = {
-    visible: true,
+    visible: true
```
Comment (javache) [inline Libraries/Modal/Modal.js]: nit: keep the comma, it keeps blame accurate

File: Libraries/Modal/Modal.js [hunk 5]
```diff
@@ -55,9 +60,18 @@ class Modal extends React.Component {
       backgroundColor: this.props.transparent ? 'transparent' : 'white',
     };
 
+    let animationType = this.props.animationType;
+    if (!animationType) {
+      // manually setting default prop here to keep support for the deprecated 'animated' prop
+      animationType = 'none';
+      if (this.props.animated) {
+        animationType = this.props.animated ? 'slide' : 'none';
```
Comment (javache) [inline Libraries/Modal/Modal.js]: nit: you've also checked the truthiness of `this.props.animated`, you can immediately set it to slide here.

File: Libraries/Modal/Modal.js [hunk 6]
```diff
@@ -77,6 +91,22 @@ class Modal extends React.Component {
...
+  animationType: PropTypes.oneOf(['none', 'slide', 'fade']),
+  transparent: PropTypes.bool,
+  visible: PropTypes.bool,
+  onRequestClose: Platform.OS === 'android' ? PropTypes.func.isRequired : PropTypes.func,
+  onShow: PropTypes.func,
+};
+
+Modal.defaultProps = {
+  visible: true
+};
```
Comment (javache) [inline Libraries/Modal/Modal.js]: Why do we need to redefine this here, if we've already got it as static props?

File: React/Views/RCTModalHostView.h [hunk 7]
```diff
@@ -16,7 +16,7 @@
 
 @interface RCTModalHostView : UIView <RCTInvalidating>
 
-@property (nonatomic, assign, getter=isAnimated) BOOL animated;
+@property (nonatomic, strong) NSString *animationType;
```
Comment (javache) [inline React/Views/RCTModalHostView.h]: String properties should be marked as copy instead of strong.

File: React/Views/RCTModalHostView.m [hunk 8]
```diff
@@ -93,7 +93,14 @@ - (void)didMoveToWindow
 
   if (!_isPresented && self.window) {
     RCTAssert(self.reactViewController, @"Can't present modal view controller without a presenting view controller");
-    [self.reactViewController presentViewController:_modalViewController animated:self.animated completion:^{
+    _modalViewController.modalTransitionStyle = UIModalTransitionStyleCoverVertical;
```
Comment (javache) [inline React/Views/RCTModalHostView.m]: Why set this here if it's already configured below?

File: React/Views/RCTModalHostView.m [hunk 9]
```diff
@@ -82,7 +82,7 @@ - (void)removeReactSubview:(UIView *)subview
 - (void)dismissModalViewController
 {
   if (_isPresented) {
-    [_modalViewController dismissViewControllerAnimated:self.animated completion:nil];
+    [_modalViewController dismissViewControllerAnimated:([self hasAnimationType]) completion:nil];
```
Comment (javache) [inline React/Views/RCTModalHostView.m]: There's no need to wrap obj-c method calls in parentheses.

File: React/Views/RCTModalHostView.m [hunk 10]
```diff
@@ -93,7 +93,14 @@ - (void)didMoveToWindow
...
     RCTAssert(self.reactViewController, @"Can't present modal view controller without a presenting view controller");
-    [self.reactViewController presentViewController:_modalViewController animated:self.animated completion:^{
+    _modalViewController.modalTransitionStyle = UIModalTransitionStyleCoverVertical;
+    
+    if ([self.animationType isEqualToString:@"fade"]) {
+      _modalViewController.modalTransitionStyle = UIModalTransitionStyleCrossDissolve;
+    } else if ([self.animationType isEqualToString:@"slide"]) {
+      _modalViewController.modalTransitionStyle = UIModalTransitionStyleCoverVertical;
+    }
+    [self.reactViewController presentViewController:_modalViewController animated:([self hasAnimationType]) completion:^{
```
Comment (javache) [inline React/Views/RCTModalHostView.m]: There's no need to wrap obj-c method calls in parentheses.

File: React/Views/RCTModalHostView.m [hunk 11]
```diff
@@ -123,6 +130,10 @@ - (BOOL)isTransparent
   return _modalViewController.modalPresentationStyle == UIModalPresentationCustom;
 }
 
+- (BOOL)hasAnimationType {
```
Comment (javache) [inline React/Views/RCTModalHostView.m]: nit: opening brace on a new line

**Comment by javache** [comment]
Just a couple of nits / style remarks. Can you also rebase your changes and make sure CI is green?

**Comment by facebook-github-bot-4** [comment]
@jemise111 updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@jemise111 updated the pull request.

**Comment by javache** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-7** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1594709974189324/int_phab) to review.

**Pull request closed without merging**